src/jit/hwintrinsicxarch.h

```diff
@@ -0,0 +1,329 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _HW_INTRINSIC_XARCH_H_
+#define _HW_INTRINSIC_XARCH_H_
+
+#ifdef FEATURE_HW_INTRINSICS
+
+enum HWIntrinsicCategory : unsigned int
+{
+    // Simple SIMD intrinsics
+    // - take Vector128/256<T> parameters
+    // - return a Vector128/256<T>
+    // - the codegen of overloads can be determined by intrinsicID and base type of returned vector
+    HW_Category_SimpleSIMD,
+
+    // IsSupported Property
+    // - each ISA class has an "IsSupported" property
+    HW_Category_IsSupportedProperty,
+
+    // IMM intrinsics
+    // - some SIMD intrinsics requires immediate value (i.e. imm8) to generate instruction
+    HW_Category_IMM,
+
+    // Scalar intrinsics
+    // - operate over general purpose registers, like crc32, lzcnt, popcnt, etc.
+    HW_Category_Scalar,
+
+    // SIMD scalar
+    // - operate over vector registers(XMM), but just compute on the first element
+    HW_Category_SIMDScalar,
+
+    // Memory access intrinsics
+    // - e.g., Avx.Load, Avx.Store, Sse.LoadAligned
+    HW_Category_MemoryLoad,
+    HW_Category_MemoryStore,
+
+    // Helper intrinsics
+    // - do not directly correspond to a instruction, such as Avx.SetAllVector256
+    HW_Category_Helper,
+
+    // Special intrinsics
+    // - have to be addressed specially
+    HW_Category_Special
+};
+
+enum HWIntrinsicFlag : unsigned int
+{
+    HW_Flag_NoFlag = 0,
+
+    // Commutative
+    // - if a binary-op intrinsic is commutative (e.g., Add, Multiply), its op1 can be contained
+    HW_Flag_Commutative = 0x1,
+
+    // Full range IMM intrinsic
+    // - the immediate value is valid on the full range of imm8 (0-255)
+    HW_Flag_FullRangeIMM = 0x2,
+
+    // Generic
+    // - must throw NotSupportException if the type argument is not numeric type
+    HW_Flag_OneTypeGeneric = 0x4,
+    // Two-type Generic
+    // - the intrinsic has two type parameters
+    HW_Flag_TwoTypeGeneric = 0x8,
+
+    // NoCodeGen
+    // - should be transformed in the compiler front-end, cannot reach CodeGen
+    HW_Flag_NoCodeGen = 0x10,
+
+    // Unfixed SIMD-size
+    // - overloaded on multiple vector sizes (SIMD size in the table is unreliable)
+    HW_Flag_UnfixedSIMDSize = 0x20,
+
+    // Complex overload
+    // - the codegen of overloads cannot be determined by intrinsicID and base type
+    HW_Flag_ComplexOverloads = 0x40,
+
+    // Multi-instruction
+    // - that one intrinsic can generate multiple instructions
+    HW_Flag_MultiIns = 0x80,
+
+    // NoContainment
+    // the intrinsic cannot be contained
+    HW_Flag_NoContainment = 0x100,
```

Comment:
I would still prefer that the flags themselves be named in a positive sense, but since the queries are now all in the positive sense, I can live with this.

Reply:
Yes, that would be https://github.com/dotnet/coreclr/issues/16656, which is a much larger refactoring.

It's probably still worth doing, but it should be easier now that we have them abstracted behind lookup functions.